### PR TITLE
[bug 489745] Callee Hierarchy Support

### PR DESCRIPTION
**Pull request by akosyakov**

### CODE REVIEW

**Comment by akosyakov** [comment]
Please review.

**Comment by dhuebner** [comment]
👎 
Please also fix dependency issue in  org.eclipse.xtext.ide merged with previous PR [bug 489745] Call Hierarchy View.

See https://xtext-builds.itemis.de/jenkins/job/xtext-runtime-tests-master-oldest/918/

[0.2]Missing requirement: Xtext IDE Core 2.10.0.v201604140035 (org.eclipse.xtext.ide 2.10.0.v201604140035) requires 'bundle org.eclipse.core.runtime [3.8.0,4.0.0)' but it could not be found
[0.2]Cannot satisfy dependency:

**Comment by dhuebner** [comment]
@akosyakov I will fix the dependency issue in master

**Comment by akosyakov** [comment]
Could someone review the content? Especially new API for IReferenceFinder.

**Comment by svenefftinge** [comment]
This week is tough, but next week for sure.

File: plugins/org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/hierarchy/LocationTableLabelProvider.xtend [hunk 1]
```diff
@@ -23,13 +23,13 @@ class LocationTableLabelProvider extends BaseLabelProvider implements ITableLabe
...
 	}
 
-	override getColumnText(Object location, int columnIndex) {
-		if (location instanceof HierarchyNodeLocation) {
+	override getColumnText(Object reference, int columnIndex) {
+		if (reference instanceof HierarchyNodeReference) {
 			switch columnIndex {
 				case 0:
-					return '' + (location.lineNumber + 1)
+					return '' + (reference.lineNumber + 1)
```
Comment (spoenemann) [inline plugins/org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/hierarchy/LocationTableLabelProvider.xtend]: `Integer.toString(reference.lineNumber + 1)`

File: plugins/org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/hierarchy/DefaultHierarchyNode.xtend [hunk 2]
```diff
@@ -27,12 +27,12 @@ class DefaultHierarchyNode implements HierarchyNode {
 	IEObjectDescription element
 
 	@Accessors(PUBLIC_GETTER)
-	val locations = <HierarchyNodeLocation>newArrayList
+	val references = <HierarchyNodeReference>newArrayList
```
Comment (spoenemann) [inline plugins/org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/hierarchy/DefaultHierarchyNode.xtend]: The getter should not expose the kind of list implementation, so better add an explicit field type `List <HierarchyNodeReference>`

Comment (spoenemann) [inline plugins/org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/hierarchy/DefaultHierarchyNode.xtend]: Same for `DefaultHierarchyRoot#roots`

File: plugins/org.eclipse.xtext/src/org/eclipse/xtext/findReferences/AnyTargetURISet.xtend [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+import org.eclipse.emf.common.util.URI
+
+/**
+ * It is used to look up all references for a resource or object scope.
+ * 
+ * @author kosyakov - Initial contribution and API
+ * @since 2.10
+ */
+@Singleton
+package class AnyTargetURISet implements TargetURIs {
```
Comment (spoenemann) [inline plugins/org.eclipse.xtext/src/org/eclipse/xtext/findReferences/AnyTargetURISet.xtend]: This is not a set (in contrast to `TargetURISet`), so it should be named `AnyTargetURIs`.

File: plugins/org.eclipse.xtext/src/org/eclipse/xtext/findReferences/AnyTargetURISet.xtend [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+	override addAllURIs(Iterable<URI> uris) {
+		// do nothing
```
Comment (spoenemann) [inline plugins/org.eclipse.xtext/src/org/eclipse/xtext/findReferences/AnyTargetURISet.xtend]: Should we throw `UnsupportedOperationException` in this and the other no-op methods?

**Comment by spoenemann** [comment]
Great work! Some general remarks:

* For the sake of consistency with other interfaces in xtext.ide I would prefer the new interfaces to start with an `I`: `IHierarchyBuilder,` `ICallHierarchyBuilder`, `IHierarchyNodeLocationProvider`, etc.
* The new class `AnyTargetURISet` seems a bit strange. Consider eliminating it by replacing the usage of `TargetURIs` in `ReferenceFinder` by something more general, e.g. `Predicate<URI>`. After all, in the reference finder `TargetURIs` is only used for filtering, but its collection properties are not relevant there.

Proposals for the UI (these could be tracked in separate Bugzillas):

* Make use of styled strings, e.g. display the type of an entry in gray color in the tree viewer.
* Add a context menu to the tree viewer
  * *Open*
  * *Focus on &lt;selected element&gt;*
  * *Copy Qualified Name*
  * *Show in...*
  * *Find References*

**Comment by akosyakov** [comment]
@spoenemann Thanks for the review! Could you give a glance at it another time?
1. Agreed on interfaces' names and usage of the predicate instead of TargetURIs.
2. I flied bugs for features which you had mentioned and also to support search scope and history.

**Comment by spoenemann** [comment]
Looks good to me!
👍

**Pull request merged**